File: src/main/java/xyz/oribuin/skyblock/listener/BlockListeners.java
package xyz.oribuin.skyblock.listener;

import dev.rosewood.rosegarden.RosePlugin;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.block.SignChangeEvent;
import xyz.oribuin.skyblock.island.Island;
import xyz.oribuin.skyblock.manager.DataManager;
import xyz.oribuin.skyblock.manager.WorldManager;

public class BlockListeners implements Listener {

    private final RosePlugin plugin;
    private final DataManager manager;
    private final WorldManager worldService;

    public BlockListeners(RosePlugin plugin) {
        this.plugin = plugin;
        this.manager = this.plugin.getManager(DataManager.class);
        this.worldService = this.plugin.getManager(WorldManager.class);
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    public void onBreak(BlockBreakEvent event) {
        if (!this.worldService.isIslandWorld(event.getBlock().getLocation())) return;

        Island island = this.manager.getIsland(event.getBlock().getLocation());
        if (island == null) return;

        if (event.getPlayer().hasPermission("skyblock.island.bypass")) return;

        if (island.isMember(event.getPlayer()) || island.isTrusted(event.getPlayer()))
            return;

        event.setCancelled(true);
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    public void onPlace(BlockPlaceEvent event) {
        if (!this.worldService.isIslandWorld(event.getBlock().getLocation())) return;

        Island island = this.manager.getIsland(event.getBlock().getLocation());
        if (island == null) return;

        if (event.getPlayer().hasPermission("skyblock.island.bypass")) return;

        if (island.isMember(event.getPlayer()) || island.isTrusted(event.getPlayer()))
            return;

        event.setCancelled(true);
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    public void onSignChange(SignChangeEvent event) {
        if (!this.worldService.isIslandWorld(event.getBlock().getLocation())) return;

        Island island = this.manager.getIsland(event.getBlock().getLocation());
        if (island == null) return;

        if (event.getPlayer().hasPermission("skyblock.island.bypass")) return;

        if (island.isMember(event.getPlayer()) || island.isTrusted(event.getPlayer()))
            return;

        event.setCancelled(true);
    }

}

File: Plugin/src/main/kotlin/xyz/oribuin/skyblock/listener/BlockListeners.kt
package xyz.oribuin.skyblock.listener

import org.bukkit.entity.Mob
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.block.SignChangeEvent
import org.bukkit.event.player.PlayerInteractEntityEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.Merchant
import xyz.oribuin.skyblock.SkyblockPlugin
import xyz.oribuin.skyblock.manager.IslandManager
import xyz.oribuin.skyblock.manager.WorldManager
import xyz.oribuin.skyblock.util.getManager

class BlockListeners(private val plugin: SkyblockPlugin) : Listener {

    private val islandManager = this.plugin.getManager<IslandManager>()
    private val worldManager = this.plugin.getManager<WorldManager>()

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    fun BlockBreakEvent.onBreak() {
        if (!worldManager.isIslandWorld(this.block.world))
            return

        if (this.player.hasPermission("skyblock.island.bypass"))
            return

        this.isCancelled = true
        val island = islandManager.getIslandFromLoc(block.location) ?: return

        if (island.members.map { it.uuid }.contains(player.uniqueId) || island.trusted.contains(player.uniqueId))
            this.isCancelled = false
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    fun BlockPlaceEvent.onPlace() {
        if (!worldManager.isIslandWorld(this.block.world))
            return

        if (this.player.hasPermission("skyblock.island.bypass"))
            return

        this.isCancelled = true
        val island = islandManager.getIslandFromLoc(block.location) ?: return

        if (island.members.map { it.uuid }.contains(player.uniqueId) || island.trusted.contains(player.uniqueId))
            this.isCancelled = false
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    fun SignChangeEvent.onSignChange() {
        if (!worldManager.isIslandWorld(this.block.world))
            return

        if (this.player.hasPermission("skyblock.island.bypass"))
            return

        this.isCancelled = true
        val island = islandManager.getIslandFromLoc(block.location) ?: return

        if (island.members.map { it.uuid }.contains(player.uniqueId) || island.trusted.contains(player.uniqueId))
            this.isCancelled = false
    }

    init {
        this.plugin.server.pluginManager.registerEvents(this, this.plugin)
    }

}